[17.0][ADD] privacy_partner_to_be_forgotten-add_gdpr_module

halbtonjazz opened the pull request:
This PR adds a new module that implements GDPR compliance features for contact anonymization in Odoo.

Key features:
- Anonymizes PII in res.partner records (name, contact details, documents)
- Preserves company relationships while anonymizing individual contacts
- Handles linked res.users records
- Removes associated messages and attachments
- Adds traceability through log notes
- Supports multi-level contact hierarchies

The anonymization process:
- Replaces names with anonymized initials
- Creates unique anonymized emails with timestamps and user id
- Clears all personal contact information
- Archives the records
- Removes all chatter history
- Handles nested contact structures

Security:
- Requires specific user permissions
- Shows confirmation wizard
- Provides clear warnings about irreversible changes

Task: 4518

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Introduced a module that enables anonymization of personal data in contact records to support GDPR compliance.
  - Added a user-friendly wizard to anonymize individual or multiple contacts, including related messages, attachments, and linked user accounts.
  - Implemented access controls to restrict anonymization actions to authorized users.
  - Provided clear warnings and confirmation steps to guide users through the irreversible anonymization process.

- **Documentation**
  - Added comprehensive guides covering configuration, usage, module description, context, and contributor information.
  - Included detailed static and README documentation with badges, use cases, and credits.

- **Tests**
  - Added extensive automated tests to validate anonymization scenarios, permission checks, and correct data handling.

- **Chores**
  - Added configuration and manifest files to support module setup and deployment.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

halbtonjazz committed:
[ADD] privacy_partner_to_be_forgotten: draft module

Add initial draft of the GDPR anonymization module to comply
with the "Right to be Forgotten" for individual contact records
(res.partner).

Includes base structure, wizard logic, access control, and UI
integration for manual anonymization via action menu.

Task: 4518

halbtonjazz committed:
[REF] privacy_partner_to_be_forgotten: refactor

Improve module implementation with demo data and code refinements:
- Add comprehensive demo data with test partners and users
- Refactor anonymization logic for better maintainability
- Update security groups with dedicated category
- Improve wizard UI and warning messages
- Fix mail.channel references to discuss.channel
- Add no_reset_password context for demo users
- Clean up docstrings and code formatting

Task: 4518

halbtonjazz committed:
[IMP] privacy_partner_to_be_forgotten: formatting & context

Improve XML formatting and context handling in the module:
- Add proper XML formatting
- Include active_test=False in partner anonymization context to handle
  archived records

These changes enhance code readability and ensure proper handling of both
active and archived partners during anonymization process.

Task: 4518

halbtonjazz committed:
[ADD] privacy_partner_to_be_forgotten: docs

Add initial documentation for the module including:
- README and subpages (CONTEXT, USAGE, CONFIGURE, etc.)
- Contributors list in readme
- HTML description for Apps Store display
- Updated manifest summary for better clarity
- Added copyright headers

Task: 4518

halbtonjazz committed:
[REF] privacy_partner_to_be_forgotten: refactor + tests

Refactor anonymization logic to improve readability,
testability, and compliance with Odoo conventions.

Changes include:
- Extracted helper methods for email, messages, attachments
- Replaced datetime with fields.Datetime for consistency
- Improved logging and translation usage
- Rewrote _anonymize_partner_data into modular structure

Added extensive test coverage:
- Parent/child anonymization scenarios
- Company restriction handling
- Access rights validation
- Attachments/messages cleanup
- Multi-partner anonymization

Minor improvements:
- Removed unused demo import
- Removed optional manifest keys
- Added changelog section to docs

Task: 4518

halbtonjazz committed:
[REF] privacy_partner_to_be_forgotten: docstrings + fixes

Improve code documentation and formatting across multiple files to enhance
readability and maintainability:

- Add detailed docstrings to core anonymization methods
- Use named placeholders in translations for better clarity
- Add ensure_one() safety check to attachment domain method
- Improve code formatting and line length compliance
- Simplify default_get logic in wizard
- Enhance anonymization success message with count

Technical improvements:
- Use rf-strings for email regex patterns
- Standardize create() and search() formatting

Task: 4518

halbtonjazz committed:
[FIX] privacy_partner_to_be_forgotten: fix license URLs

Update all license header URLs to correctly point to LGPL-3.0 instead of AGPL
across the module. This ensures consistency between the declared license
(LGPL-3.0) and its reference URL.

Files updated:
- Python files (__init__.py, __manifest__.py, models, tests, wizards)
- XML files (demo data, security, views)

This change improves legal accuracy and maintains proper license
documentation throughout the codebase.

Task: 4518

halbtonjazz commented on privacy_partner_to_be_forgotten/readme/CONTEXT.md:
```diff
@@ -0,0 +1 @@
+This module aims to help companies comply with the European Union General Data Protection Regulation (GDPR) by providing a mechanism to anonymize personal data in `res.partner` records.
```
No need to add headers, because they will be generated automatically from "/readme" files.

halbtonjazz commented on privacy_partner_to_be_forgotten/wizards/__init__.py:
```diff
@@ -0,0 +1,4 @@
+# Copyright (C) 2025 Cetmix OÜ
+# License LGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
```
Done as described in the task description

halbtonjazz replied:
Good point, thanks

halbtonjazz commented on privacy_partner_to_be_forgotten/readme/USAGE.md:
```diff
@@ -0,0 +1,17 @@
+## To anonymize a partner record:
```
Done as described in the task description

halbtonjazz commented on privacy_partner_to_be_forgotten/readme/USAGE.md:
```diff
@@ -0,0 +1,17 @@
+3. Go to **Actions** → **Anonymize (GDPR)**.
+4. A confirmation wizard will appear. Press the **"Confirm"** button to proceed.
+
+Once you confirm, Odoo will anonymize the partner's record and all related child records, including:
+- Chatter messages and attachments
+- Linked `res.users` (user) records, if applicable
+
```
Done as described in the task description

halbtonjazz commented on privacy_partner_to_be_forgotten/security/security.xml:
```diff
@@ -0,0 +1,23 @@
+    </record>
+
+    <!-- Define the group for Anonymize Contacts -->
+    <record id="group_partner_anonymize" model="res.groups">
+        <field name="name">Administrator</field>
+        <field name="category_id" ref="module_category_anonymize_contacts" />
+        <field name="users" eval="[(4, ref('base.user_admin'))]" />
+        <field
+            name="comment"
+        >Users in this group can anonymize contact information for GDPR compliance.</field>
+    </record>
```
I propose to leave the current implementation

halbtonjazz commented on privacy_partner_to_be_forgotten/security/security.xml:
```diff
@@ -0,0 +1,23 @@
+-->
+<odoo>
+    <!-- Create a new category for Anonymize Contacts under Administration -->
+    <record model="ir.module.category" id="module_category_anonymize_contacts">
+        <field name="name">Anonymize Contacts</field>
+        <field name="parent_id" ref="base.module_category_administration" />
+        <field name="sequence">101</field>
+    </record>
```
Made similar to addons/oca-data-protection/privacy/security/data_protection.xml

halbtonjazz commented on privacy_partner_to_be_forgotten/readme/DESCRIPTION.md:
```diff
@@ -0,0 +1,21 @@
+The module allows anonymizing contacts that represent natural persons (i.e., individuals) when they are no longer needed or when a data erasure request is received.
```
No need to add headers, because they will be generated automatically from "/readme" files.

halbtonjazz commented on privacy_partner_to_be_forgotten/readme/DESCRIPTION.md:
```diff
@@ -0,0 +1,21 @@
+The module allows anonymizing contacts that represent natural persons (i.e., individuals) when they are no longer needed or when a data erasure request is received.
+
+This module provides a method to anonymize all personally identifiable information (PII) from a `res.partner` record:
+- Name (e.g., replaced with initials or a generic label) for individual contacts
```
Done as described in the task description

halbtonjazz commented on privacy_partner_to_be_forgotten/readme/DESCRIPTION.md:
```diff
@@ -0,0 +1,21 @@
+
+Additionally, the module removes all chatter messages (`mail.message`) and attachments associated with the partner record, as these may contain personal information, such as communication history, internal notes, or file uploads. This ensures complete anonymization and supports full GDPR compliance.
+
+If the partner is linked to one or more `res.users` records (e.g., as a portal user or employee), the module will anonymize these user records:
+-  The user login and email are replaced with anonymized values, and the user is archived.
```
Done as described in the task description

halbtonjazz commented on privacy_partner_to_be_forgotten/wizards/partner_anonymize_wizard_views.xml:
```diff
@@ -0,0 +1,62 @@
+        <field name="model">partner.anonymize.wizard</field>
+        <field name="arch" type="xml">
+            <form string="Anonymize Partners">
+                <div class="alert alert-warning" role="alert">
+                    <h3>Confirmation Required: Anonymization Process</h3>
+                    <p
+                    >Once you confirm, Odoo will anonymize the selected partner(s) and all related records, including:
+                        <ul>
+                            <li>Personal information (name, email, phone, address)</li>
+                            <li
+                            >User account information (if the contact has system access)</li>
+                            <li>Messages and attachments in chat history</li>
+                            <li>Contact's activity history</li>
+                        </ul>
+                    </p>
+                    <p><strong>Important:</strong></p>
+                    <ul>
+                        <li>The anonymization operation <strong
+                            >cannot be undone</strong>.</li>
+                        <li>Once the data is anonymized, it <strong
+                            >cannot be restored</strong>. Please proceed carefully.</li>
+                        <li>Company records cannot be anonymized.</li>
+                        <li>You <strong
+                            >carry full responsibility</strong> for using the module's features in compliance with GDPR regulations.</li>
+                    </ul>
+                </div>
```
@geomer198
is it worth taking?

halbtonjazz commented on privacy_partner_to_be_forgotten/models/res_partner.py:
```diff
@@ -0,0 +1,206 @@
+            "signature": False,
+        }
+
+    def _anonymize_user(self, anonymized_email):
+        """Anonymize related user accounts.
+
+        Args:
+            anonymized_email (str): Generated anonymous email to be set
+                for user accounts
+        """
+        if self.user_ids:
+            self.user_ids.write(self._prepare_user_anonymized_vals(anonymized_email))
+            self.user_ids = self.env["res.users"].browse(self.user_ids.ids)
+
```
I removed that piece

halbtonjazz commented on privacy_partner_to_be_forgotten/models/res_partner.py:
```diff
@@ -0,0 +1,206 @@
+        initials = "".join(part[0].upper() for part in name_parts if part)
+        return initials or "X"
+
+    def _generate_anonymized_email(self, initials):
+        """Generate anonymized email address.
+
+        Args:
+            initials (str): Partner name initials (e.g., "JD" for "John Doe")
+
+        Returns:
+            str: Anonymized email address in format
+                "initials_dd.mm.yy_id@anonymized.oca"
+        """
+        date_stamp = fields.Date.today().strftime("%d.%m.%y")
+        return f"{initials.lower()}_{date_stamp}_{self.id}@anonymized.oca"
+
```
Done as described in the task description

halbtonjazz commented on privacy_partner_to_be_forgotten/models/res_partner.py:
```diff
@@ -0,0 +1,206 @@
+            "active": False,
+        }
+
+    def _prepare_user_anonymized_vals(self, anonymized_email):
+        """Prepare values for anonymizing related user accounts.
+
+        Args:
+            anonymized_email (str): Generated anonymous email
+                (format: "xx_dd.mm.yy_id@anonymized.oca")
+
+        Returns:
+            dict: Dictionary containing user fields to update
+                (login, email, active status, signature)
+        """
+        return {
+            "login": anonymized_email,
+            "email": anonymized_email,
+            "active": False,
+            "signature": False,
+        }
```
I propose to leave the current implementation. In this case the signature is auto-generated by Odoo when name/email/login is changed, so tests still pass even if we set it to False explicitly.

halbtonjazz commented on privacy_partner_to_be_forgotten/wizards/partner_anonymize_wizard.py:
```diff
@@ -0,0 +1,73 @@
+                % anonymized_names
+            )
+
+    def action_confirm(self):
+        """Confirm and process partner anonymization for multiple partners"""
+        self.ensure_one()
+        self._validate_partners_for_anonymization()
+
+        for partner in self.partner_ids:
+            partner.anonymize_partner_data()
+
```
This is a complication. I suggest leaving the current implementation

halbtonjazz commented on privacy_partner_to_be_forgotten/tests/test_partner_anonymize.py:
```diff
@@ -0,0 +1,415 @@
+            }
+        )
+
+    def _create_anonymize_wizard(self, partner_ids, user):
+        """Helper method to create and return an anonymization wizard"""
+        with Form(
+            self.env["partner.anonymize.wizard"]
+            .with_user(user)
+            .with_context(default_partner_ids=partner_ids, active_ids=partner_ids)
+        ) as wizard:
+            wizard_id = wizard.save()
+
+        return wizard_id
+
```
I suggest leaving the current implementation

halbtonjazz commented on privacy_partner_to_be_forgotten/README.rst:
```diff
@@ -0,0 +1,171 @@
+- You carry full responsibility for using the module's features, in line
+  with GDPR compliance.
+
+Changelog
+=========
+
+
+
+Bug Tracker
+===========
+
```
Not required readme is generated automatically, maybe I will remove the newsfragments folder

geomer198 commented on privacy_partner_to_be_forgotten/demo/demo_partner_anonymize_data.xml:
```diff
@@ -0,0 +1,142 @@
+    Copyright (C) 2025 Cetmix OÜ
+    License LGPL-3.0 or later (https://www.gnu.org/licenses/lgpl-3.0).
+-->
+<odoo>
```
Please add `noupdate=1`

halbtonjazz replied:
Thanks, I've added it

geomer198 commented on privacy_partner_to_be_forgotten/__manifest__.py:
```diff
@@ -0,0 +1,21 @@
+    "development_status": "Beta",
+    "category": "Data Protection",
+    "summary": "Anonymize partner data for GDPR compliance",
+    "author": "Cetmix OÜ, Odoo Community Association (OCA)",
```
Author just `Cetmix, Odoo Community Association (OCA)`

halbtonjazz replied:
fixed, thanks

geomer198 commented on privacy_partner_to_be_forgotten/demo/demo_partner_anonymize_data.xml:
```diff
@@ -0,0 +1,142 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<!--
```
Please remove all `Copyright` string from xml files.

halbtonjazz replied:
Done

geomer198 reviewed (approved):
LGTM!

halbtonjazz committed:
[REF] privacy_partner_to_be_forgotten: refactor test & cleanup

Improve module structure and maintainability through several changes:
- Replace XML demo data with programmatic test data creation
- Remove redundant browse call in user anonymization
- Add noupdate flag to demo data
- Remove copyright headers from XML files
- Update company name from "Cetmix OÜ" to "Cetmix"
- Fix license reference consistency (AGPL -> LGPL-3.0)
These changes make the module more maintainable and ensure consistent
naming and licensing across all files.

Task: 4518

halbtonjazz commented:
Closing this PR to reopen it from the cetmix/oca-data-protection repository to the OCA/data-protection to allow Runbot to build the module and enable proper functional testing.
